File: backend/internal/services/category_service.go
package services

import (
	"errors"
	"github.com/huy1235588/fashion-e-commerce/internal/models"
	"github.com/huy1235588/fashion-e-commerce/internal/repositories"
	"gorm.io/gorm"
)

// CategoryService handles category business logic
type CategoryService struct {
	repo repositories.CategoryRepository
}

// NewCategoryService creates a new category service
func NewCategoryService(repo repositories.CategoryRepository) *CategoryService {
	return &CategoryService{repo: repo}
}

// CreateCategory creates a new category
func (s *CategoryService) CreateCategory(category *models.Category) error {
	// Check if slug already exists
	existing, err := s.repo.FindBySlug(category.Slug)
	if err != nil && !errors.Is(err, gorm.ErrRecordNotFound) {
		return err
	}
	if existing != nil {
		return errors.New("category with this slug already exists")
	}

	return s.repo.Create(category)
}

// GetCategory retrieves a category by ID
func (s *CategoryService) GetCategory(id uint) (*models.Category, error) {
	return s.repo.FindByID(id)
}

// GetCategoryBySlug retrieves a category by slug
func (s *CategoryService) GetCategoryBySlug(slug string) (*models.Category, error) {
	return s.repo.FindBySlug(slug)
}

// UpdateCategory updates a category
func (s *CategoryService) UpdateCategory(id uint, updates *models.Category) error {
	category, err := s.repo.FindByID(id)
	if err != nil {
		return err
	}

	// Check if slug is being changed and if it conflicts
	if updates.Slug != category.Slug {
		existing, err := s.repo.FindBySlug(updates.Slug)
		if err != nil && !errors.Is(err, gorm.ErrRecordNotFound) {
			return err
		}
		if existing != nil && existing.ID != id {
			return errors.New("category with this slug already exists")
		}
	}

	category.Name = updates.Name
	category.Description = updates.Description
	category.Slug = updates.Slug

	return s.repo.Update(category)
}

// DeleteCategory deletes a category
func (s *CategoryService) DeleteCategory(id uint) error {
	return s.repo.Delete(id)
}

// ListCategories retrieves all categories
func (s *CategoryService) ListCategories() ([]models.Category, error) {
	return s.repo.List()
}

File: frontend/src/services/category.service.ts
import { apiClient } from '@/lib/api';
import { API_ENDPOINTS } from '@/lib/constants';
import type { Category } from '@/types';

export interface CreateCategoryInput {
    name: string;
    description?: string;
    slug: string;
}

export interface UpdateCategoryInput {
    name?: string;
    description?: string;
    slug?: string;
}

class CategoryService {
    // Get all categories (public)
    async getCategories(): Promise<Category[]> {
        const response = await apiClient.get<{ data: Category[] }>(API_ENDPOINTS.CATEGORIES);
        return response.data;
    }

    // Get category by ID
    async getCategoryById(id: number): Promise<Category> {
        const response = await apiClient.get<{ data: Category }>(API_ENDPOINTS.CATEGORY_DETAIL(id));
        return response.data;
    }

    // Create category (admin)
    async createCategory(data: CreateCategoryInput): Promise<Category> {
        const response = await apiClient.post<{ data: Category }>(
            API_ENDPOINTS.ADMIN_CATEGORIES,
            data
        );
        return response.data;
    }

    // Update category (admin)
    async updateCategory(id: number, data: UpdateCategoryInput): Promise<Category> {
        const response = await apiClient.put<{ data: Category }>(
            `${API_ENDPOINTS.ADMIN_CATEGORIES}/${id}`,
            data
        );
        return response.data;
    }

    // Delete category (admin)
    async deleteCategory(id: number): Promise<void> {
        await apiClient.delete(`${API_ENDPOINTS.ADMIN_CATEGORIES}/${id}`);
    }
}

export const categoryService = new CategoryService();
